[Model element] Support `get/setEntityTransform` for model element

mwyrzykowski opened the pull request:
#### ba78fd40d31d11aae50a97834aecd6a396aec73c
<pre>
[Model element] Support `get/setEntityTransform` for model element
<a href="https://bugs.webkit.org/show_bug.cgi?id=299973">https://bugs.webkit.org/show_bug.cgi?id=299973</a>
<a href="https://rdar.apple.com/161754835">rdar://161754835</a>

Reviewed by Etienne Segonzac.

Pass 4x4 transform from JS to GPU process. Assume identity after the model is
loaded.

Test: model-element/model-element-entity-transform.html passes locally after this change

* Source/WebCore/Modules/Model/Implementation/DDMeshImpl.cpp:
(WebCore::DDModel::DDMeshImpl::setEntityTransform):
(WebCore::DDModel::DDMeshImpl::entityTransform const):
* Source/WebCore/Modules/Model/Implementation/DDMeshImpl.h:
* Source/WebCore/Modules/Model/InternalAPI/DDMesh.h:
* Source/WebCore/Modules/model-element/DDModelPlayer.h:
* Source/WebCore/Modules/model-element/DDModelPlayer.mm:
(WebCore::areSameSignAndAlmostEqual):
(WebCore::makeTransformMatrix):
(WebCore::DDModelPlayer::supportsTransform):
(WebCore::DDModelPlayer::entityTransform const):
(WebCore::DDModelPlayer::setEntityTransform):
* Source/WebGPU/WebGPU/DDMesh.h:
* Source/WebGPU/WebGPU/DDMesh.mm:
(WebGPU::DDMesh::setTransform):
(wgpuDDMeshSetTransform):
* Source/WebGPU/WebGPU/DDModelTypes.h:
* Source/WebGPU/WebGPU/UsdModelRenderer.swift:
(DDBridgeReceiver.setTransform(_:)):
* Source/WebGPU/WebGPU/WebGPUExt.h:
* Source/WebKit/GPUProcess/graphics/Model/RemoteDDMesh.cpp:
(WebKit::RemoteDDMesh::updateTransform):
* Source/WebKit/GPUProcess/graphics/Model/RemoteDDMesh.h:
* Source/WebKit/GPUProcess/graphics/Model/RemoteDDMesh.messages.in:
* Source/WebKit/WebProcess/GPU/graphics/Model/RemoteDDMeshProxy.cpp:
(WebKit::DDModel::RemoteDDMeshProxy::update):
(WebKit::DDModel::RemoteDDMeshProxy::setEntityTransform):
(WebKit::DDModel::RemoteDDMeshProxy::entityTransform const):
* Source/WebKit/WebProcess/GPU/graphics/Model/RemoteDDMeshProxy.h:

Canonical link: <a href="https://commits.webkit.org/302424@main">https://commits.webkit.org/302424@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/123ae4a76a5a0f728efe48e4b7ce6865cae7132f

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows | Apple Internal |
| ----- | ---------------------- | ------- |  ----- |  --------- | ------ |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/129028 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/159/builds/1287 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/39864 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/136409 "Built successfully") | [❌ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/80385 "Hash 123ae4a7 for PR 52695 does not build (failure)") | [✅ 🛠 ios-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/f2c6d2ed-bd8b-4d70-a499-64172ae81bd0/b6540997-50e7-41d8-a93b-719d8f3aeaa0) 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/130899 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/155/builds/1219 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/156/builds/1164 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/98237 "Passed tests") | [❌ 🧪 win-tests](https://ews-build.webkit.org/#/builders/59/builds/80385 "Hash 123ae4a7 for PR 52695 does not build (failure)") | [✅ 🛠 mac-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/f2c6d2ed-bd8b-4d70-a499-64172ae81bd0/b24f23b7-8f48-49aa-bfc5-374e3753b009) 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/131975 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/162/builds/938 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/115582 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/78881 "Passed tests") | | [✅ 🛠 vision-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/f2c6d2ed-bd8b-4d70-a499-64172ae81bd0/d198219d-3b8b-40ee-a460-89e31e34bf5e) 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/154/builds/866 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/33697 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/79688 "Built successfully") | | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/20/builds/121020 "Built successfully and passed tests") | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/109306 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/34195 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/138883 "Built successfully") | | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/127478 "Built successfully and passed tests") | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/153/builds/1082 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/161/builds/1056 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/106775 "Passed tests") | | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/160/builds/1139 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/111918 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/106601 "Passed tests") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/27143 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/164/builds/894 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/30442 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/53587 "Built successfully") | | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/158/builds/1156 "Built successfully") | [❌ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/64496 "Failed to build and analyze WebKit") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/160495 "Built successfully") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/157/builds/990 "Built successfully") | | [✅ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/25/builds/40058 "Passed tests") | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/163/builds/1039 "Built successfully") | | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/152/builds/1083 "Built successfully") | | | | 
<!--EWS-Status-Bubble-End-->

weinig commented on Source/WebCore/Modules/model-element/HTMLModelElement.h:
```diff
@@ -163,8 +163,10 @@ class HTMLModelElement final : public HTMLElement, private CachedRawResourceClie
 
-#if ENABLE(MODEL_ELEMENT_STAGE_MODE_INTERACTION)
+#if ENABLE(MODEL_ELEMENT_STAGE_MODE)
     WEBCORE_EXPORT bool supportsStageModeInteraction() const;
```
Seems pretty weird for a function with the text "StageModeInteraction" to not be in the ENABLE(MODEL_ELEMENT_STAGE_MODE_INTERACTION) block.

mwyrzykowski replied:
Oh good observation, I will rename this member function

etiennesegonzac commented on Source/WebKit/GPUProcess/graphics/Model/RemoteDDMesh.cpp:
```diff
@@ -143,6 +143,15 @@ void RemoteDDMesh::updateMaterial(const WebCore::DDModel::DDUpdateMaterialDescri
+#if PLATFORM(COCOA)
+    m_backing->setEntityTransform(transform);
+#else
+    UNUSED_PARAM(descriptor);
```
transform?

mwyrzykowski replied:
Will remove the `PLATFORM(COCOA)` checks because the entire file is in `ENABLE(GPU_PROCESS_MODEL)` which is why this didn't break a non-Cocoa build

etiennesegonzac commented on Source/WebGPU/WebGPU/UsdModelRenderer.swift:
```diff
@@ -1541,4 +1541,15 @@ extension DDBridgeReceiver {
+    @objc(setTransform:)
+    func setTransform(_ transform: simd_float4x4) {
+        #if canImport(DirectDrawBackend)
+        for meshArray in meshInstances {
```
nit: `for meshArray in meshInstances.values` and no `.value` below

etiennesegonzac commented on Source/WebCore/Modules/model-element/DDModelPlayer.mm:
```diff
@@ -336,6 +338,54 @@ void setOpaque(bool opaque)
+
+    if (RefPtr model = m_currentModel) {
+        auto [center, extents] = m_currentModel->getCenterAndExtents();
+        simd_float4x4 result = matrix_identity_float4x4;
```
Maybe for a follow-up, but I think the ModelProcessPlayer keeps the current rotation in this case.

mwyrzykowski replied:
I can fix that here, it's pretty simple since these are all RST matrices

etiennesegonzac commented on Source/WebCore/Modules/model-element/DDModelPlayer.mm:
```diff
@@ -336,6 +338,54 @@ void setOpaque(bool opaque)
+    if (RefPtr model = m_currentModel) {
+        auto [center, extents] = m_currentModel->getCenterAndExtents();
+        simd_float4x4 result = matrix_identity_float4x4;
+        float scale = simd_length(extents) * .5f;
```
Not entirely sure I follow how this gives us the "will fit into the layer".

mwyrzykowski replied:
Indeed I think this is incorrect, will fix that

mwyrzykowski replied:
(still correcting this, should be fixed in next PR update)

mwyrzykowski replied:
This is resolved now

etiennesegonzac commented on Source/WebKit/WebProcess/GPU/graphics/Model/RemoteDDMeshProxy.cpp:
```diff
@@ -36,13 +36,38 @@
+    return std::abs(absA - absB) < tolerance * std::min(absA, absB);
+}
+
+static simd_float4x4 makeTransformMatrix(
```
should we move/inline these somewhere so the DDModelPlayer can use them?

mwyrzykowski replied:
I removed the usage in DDModelPlayer so I propose holding off until its used elsewhere

etiennesegonzac reviewed (commented):
Note: you have a random
```

i
```
at the end for your commit message :)

etiennesegonzac commented on Source/WebCore/Modules/model-element/DDModelPlayer.mm:
```diff
@@ -181,6 +181,14 @@ void setOpaque(bool opaque)
                 client->didFinishLoading(protectedThis.get());
                 auto [simdCenter, simdExtents] = protectedThis->m_currentModel->getCenterAndExtents();
                 client->didUpdateBoundingBox(protectedThis.get(), FloatPoint3D(simdCenter.x, simdCenter.y, simdCenter.z), FloatPoint3D(simdExtents.x, simdExtents.y, simdExtents.z));
+                if (auto entityTransform = protectedThis->m_currentModel->entityTransform()) {
```
can we extract this in a `DDModelPlayer::entityTransformDidChange()` or similar (that we can re-use in `setStageMode`)

etiennesegonzac commented on Source/WebCore/Modules/model-element/DDModelPlayer.mm:
```diff
@@ -215,8 +223,9 @@ void setOpaque(bool opaque)
-void DDModelPlayer::sizeDidChange(LayoutSize)
+void DDModelPlayer::sizeDidChange(LayoutSize layoutSize)
 {
+    m_currentScale = static_cast<float>(layoutSize.width());
```
Discussed offline, but leaving a note here: what about height :)

etiennesegonzac replied:
`layoutSize.minDimension()`?

etiennesegonzac commented on Source/WebKit/WebProcess/GPU/graphics/Model/RemoteDDMeshProxy.cpp:
```diff
@@ -104,14 +132,33 @@ void RemoteDDMeshProxy::addMesh(const WebCore::DDModel::DDMeshDescriptor& descri
+    m_maxCorner = simd_max(maxCorner, m_maxCorner);
+
+    auto [center, extents] = getCenterAndExtents();
+    setCameraDistance(std::max(extents.x, extents.y) * .5f);
```
Just double checking: this is stable while we update the rotation during orbiting right?

mwyrzykowski replied:
It should be as orbit mode is handled by adjusting the model’s scale

mwyrzykowski committed:
[Model element] Support `get/setEntityTransform` for model element
https://bugs.webkit.org/show_bug.cgi?id=299973
rdar://161754835

Reviewed by Etienne Segonzac.

Pass 4x4 transform from JS to GPU process. Assume identity after the model is
loaded.

Test: model-element/model-element-entity-transform.html passes locally after this change

* Source/WebCore/Modules/Model/Implementation/DDMeshImpl.cpp:
(WebCore::DDModel::DDMeshImpl::setEntityTransform):
(WebCore::DDModel::DDMeshImpl::entityTransform const):
* Source/WebCore/Modules/Model/Implementation/DDMeshImpl.h:
* Source/WebCore/Modules/Model/InternalAPI/DDMesh.h:
* Source/WebCore/Modules/model-element/DDModelPlayer.h:
* Source/WebCore/Modules/model-element/DDModelPlayer.mm:
(WebCore::areSameSignAndAlmostEqual):
(WebCore::makeTransformMatrix):
(WebCore::DDModelPlayer::supportsTransform):
(WebCore::DDModelPlayer::entityTransform const):
(WebCore::DDModelPlayer::setEntityTransform):
* Source/WebGPU/WebGPU/DDMesh.h:
* Source/WebGPU/WebGPU/DDMesh.mm:
(WebGPU::DDMesh::setTransform):
(wgpuDDMeshSetTransform):
* Source/WebGPU/WebGPU/DDModelTypes.h:
* Source/WebGPU/WebGPU/UsdModelRenderer.swift:
(DDBridgeReceiver.setTransform(_:)):
* Source/WebGPU/WebGPU/WebGPUExt.h:
* Source/WebKit/GPUProcess/graphics/Model/RemoteDDMesh.cpp:
(WebKit::RemoteDDMesh::updateTransform):
* Source/WebKit/GPUProcess/graphics/Model/RemoteDDMesh.h:
* Source/WebKit/GPUProcess/graphics/Model/RemoteDDMesh.messages.in:
* Source/WebKit/WebProcess/GPU/graphics/Model/RemoteDDMeshProxy.cpp:
(WebKit::DDModel::RemoteDDMeshProxy::update):
(WebKit::DDModel::RemoteDDMeshProxy::setEntityTransform):
(WebKit::DDModel::RemoteDDMeshProxy::entityTransform const):
* Source/WebKit/WebProcess/GPU/graphics/Model/RemoteDDMeshProxy.h:

Canonical link: https://commits.webkit.org/302424@main
